go/private/sdk.bzl

```diff
@@ -142,73 +142,164 @@ _go_download_sdk = repository_rule(
     pv = _parse_version(version)
     if pv == None or len(pv) < 3:
         fail("error parsing sdk version: " + version)
     major, minor, patch = pv[0], pv[1], pv[2]
     prerelease = pv[3] if len(pv) > 3 else ""
     return """
-MAJOR_VERSION = "{major}"
-MINOR_VERSION = "{minor}"
-PATCH_VERSION = "{patch}"
-PRERELEASE_SUFFIX = "{prerelease}"
+{prefix}MAJOR_VERSION = "{major}"
+{prefix}MINOR_VERSION = "{minor}"
+{prefix}PATCH_VERSION = "{patch}"
+{prefix}PRERELEASE_SUFFIX = "{prerelease}"
 """.format(
+        prefix = prefix,
         major = major,
         minor = minor,
         patch = patch,
         prerelease = prerelease,
     )
 
-def _go_toolchains_impl(ctx):
-    if not ctx.attr.goos and not ctx.attr.goarch:
+def _to_constant_name(s):
+    return "".join([c if c.isalnum() else "_" for c in s.elems()]).upper()
+
+def _go_toolchains_single_definition(ctx, *, prefix, goos, goarch, sdk_repo, sdk_type, sdk_version):
+    if not goos and not goarch:
         goos, goarch = _detect_host_platform(ctx)
     else:
-        if not ctx.attr.goos:
+        if not goos:
             fail("goarch set but goos not set")
-        if not ctx.attr.goarch:
+        if not goarch:
             fail("goos set but goarch not set")
-        goos, goarch = ctx.attr.goos, ctx.attr.goarch
 
-    sdk_repo = ctx.attr.sdk_repo
-    sdk_type = ctx.attr.sdk_type
+    chunks = []
+    loads = []
+    identifier_prefix = _to_constant_name(prefix)
 
     # If a sdk_version attribute is provided, use that version. This avoids
     # eagerly fetching the SDK repository. But if it's not provided, we have
     # no choice and must load version constants from the version.bzl file that
     # _sdk_build_file creates. This will trigger an eager fetch.
-    version = ctx.attr.sdk_version
-    if version:
-        version_constants = _define_version_constants(version)
+    if sdk_version:
+        chunks.append(_define_version_constants(sdk_version, prefix = identifier_prefix))
     else:
-        version_constants = 'load("@{}//:version.bzl", "MAJOR_VERSION", "MINOR_VERSION", "PATCH_VERSION", "PRERELEASE_SUFFIX")'.format(sdk_repo)
+        loads.append("""load(
+    "@{sdk_repo}//:version.bzl",
+    {identifier_prefix}MAJOR_VERSION = "MAJOR_VERSION",
+    {identifier_prefix}MINOR_VERSION = "MINOR_VERSION",
+    {identifier_prefix}PATCH_VERSION = "PATCH_VERSION",
+    {identifier_prefix}PRERELEASE_SUFFIX = "PRERELEASE_SUFFIX",
+)""".format(
+            sdk_repo = sdk_repo,
+            identifier_prefix = identifier_prefix,
+        ))
+
+    chunks.append("""declare_bazel_toolchains(
+    prefix = "{prefix}",
+    go_toolchain_repo = "@{sdk_repo}",
+    host_goarch = "{goarch}",
+    host_goos = "{goos}",
+    major = {identifier_prefix}MAJOR_VERSION,
+    minor = {identifier_prefix}MINOR_VERSION,
+    patch = {identifier_prefix}PATCH_VERSION,
+    prerelease = {identifier_prefix}PRERELEASE_SUFFIX,
+    sdk_type = "{sdk_type}",
+)
```

Comment:
I think it would be good to either have a test for this function, to prevent changes from causing unintended generated `declare_bazel_toolchains(` rules.

Reply:
Added tests for the two `sdk_version` cases.